lucene/core/src/java/org/apache/lucene/codecs/lucene80/Lucene80Codec.java

```diff
@@ -91,7 +91,11 @@ public Lucene80Codec() {
    *             flushed/merged segments.
    */
   public Lucene80Codec(Mode mode) {
-    super("Lucene80");
+    this("Lucene80", mode);
+  }
+  
+  protected Lucene80Codec(String name, Mode mode) {
+    super(name);
```

Comment:
can you use FilterCodec instead?

Reply:
> Use FilterCodec instead?
Yes! Good point. I'll do that.

> Use Lucene50PostingsWriter instead?
DeltaBaseTermStateSerializer writes each TermState with file pointers delta encoded relatively to a base pointer (which is the block base file pointer). And it does only that thing. This is different from Lucene50PostingsWriter.encodeTerm which writes the TermState with file pointers delta encoded relatively to the previous TermState written.
This logic to delta encode relatively to the base pointer allows us to read TermStates in random access, no sequential reading required. This allows us to make a binary search inside the block itself and read only one TermState.
To summarize: No, we cannot use Lucene50PostingsWriter.encodeTerm. That said DeltaBaseTermStateSerializer could be located elsewhere, the only issue is that it writes IntBlockTermState to stay compatible and still use Lucene50PostingsReader.postings() which requires IntBlockTermState because it casts internally the provided BlockTermState.

Reply:
Has this binary search within blocks proved useful? In general I don't have objections to adding new file formats, but I'm a bit worried when they use implementation details of existing file formats instead of using APIs, as this might make these other formats harder to evolve.

Reply:
Two points here:

The binary search inside the block allowed the posting format to improve the seek performance significantly without increasing the size of the FST. Indeed we can have larger blocks for same seek speed, so the FST size is smaller since there are less leaves.

We do use the API for Lucene50PostingsReader.postings(). This is a public method, but unfortunately it internally casts the TermState param to an internal IntBlockTermState. Maybe it would be cleaner for this API to have a public interface for IntBlockTermState, or make it public since its fields are accessed directly. That way we wouldn't have to have a class in this package.

Reply:
> Maybe it would be cleaner for this API to have a public interface for IntBlockTermState, or make it public since its fields are accessed directly.

+1!

Reply:
@jpountz about (private) IntBlockTermState cast in public Lucene50PostingsReader.postings(), should I create a separate JIRA issue? I think it deserves it.

Reply:
I have created LUCENE-8906 (Lucene50PostingsReader.postings() casts BlockTermState param to private IntBlockTermState) to discuss this specific issue separately.